File: evm/src/observer/curation/exhibit.ts
import { Chain, PrismaPromise } from "@prisma/client";
import {
  CHAIN,
  CONTRACT_CURATION,
  EVENTOFFSET_ID,
  KEY_COLLECTION_TOKENS,
  randomProvider,
} from "../../config";
import { prisma } from "../../io";
import { ERC721__factory } from "../../typechain";
import { TypedEvent } from "../../typechain/common";
import { ExhibitChangedEvent } from "../../typechain/Curation";
import { State } from "../../types";
import { handleError, Observer } from "../observer";
import { redis } from "../../io/redis";

const eventTypeToStatus = {
  ExhibitListed: "listing",
  ExhibitCanceled: "reserved",
  ExhibitSold: "sold",
  ExhibitFrozen: "frozen",
  ExhibitRedeemed: "redeemed",
};
export class ExhibitObserver extends Observer {
  async processAll<T extends TypedEvent>(
    state: State,
    events: T[],
    startBlockNo: number,
    endBlockNo: number
  ): Promise<State> {
    const newEvents = events.filter(
      (e) => e.blockNumber > Number(state.exhibit_excuted_offset)
    );
    return super.processAll(state, newEvents, startBlockNo, endBlockNo);
  }
  async process<T extends TypedEvent>(
    state: State,
    event: T
  ): Promise<{ success: boolean; state: State }> {
    const blockNo = BigInt(event.blockNumber);
    const [
      id,
      eventType,
      galleryId,
      collection,
      tokenId,
      origin,
      expiration,
      price,
      _,
      url,
      detail,
      timestamp,
    ] = (event as ExhibitChangedEvent).args;
    const updatedAt = new Date(timestamp.toNumber() * 1000);
    const exhibit = await prisma.curationExhibit.findFirst({
      where: {
        index: id.toBigInt(),
        root: CONTRACT_CURATION,
        status: {
          in: ["listing", "reserved"],
        },
      },
      orderBy: {
        updatedAt: "desc",
      },
    });
    const updateExhibit = prisma.curationExhibit.upsert({
      where: {
        id: exhibit?.id || "",
      },
      create: {
        index: id.toBigInt(),
        chain: CHAIN as Chain,
        root: CONTRACT_CURATION,
        galleryIndex: galleryId.toBigInt(),
        curator: origin,
        collectionIdentifier: collection,
        tokenCreator: "",
        tokenIdentifier: tokenId.toString(),
        propertyVersion: 0,
        origin,
        price: price.toString(),
        currency: "0x0000000000000000000000000000000000000000",
        decimals: 18,
        expiredAt: new Date(expiration.toNumber() * 1000),
        location: "",
        url: url,
        detail: detail,
        status: eventTypeToStatus[eventType],
        updatedAt,
      },
      update: {
        status: eventTypeToStatus[eventType],
        price: price.toString(),
        galleryIndex: galleryId.toBigInt(),
        origin,
        curator: origin,
        detail,
      },
    });
    const updateOffset = prisma.eventOffset.update({
      where: {
        id: +EVENTOFFSET_ID,
      },
      data: {
        exhibit_excuted_offset: blockNo,
      },
    });
    let owner: string;
    let txs: PrismaPromise<any>[] = [updateExhibit, updateOffset];
    if (eventType == "ExhibitSold") {
      owner = await ERC721__factory.connect(
        collection,
        randomProvider()
      ).ownerOf(tokenId);
      const updateTransaction = prisma.transaction.upsert({
        where: {
          chain_tokenId_collectionId_txType_txTimestamp: {
            chain: CHAIN as Chain,
            tokenId: tokenId.toString(),
            collectionId: collection,
            txType: "SALE",
            txTimestamp: updatedAt,
          },
        },
        create: {
          chain: CHAIN as Chain,
          tokenId: tokenId.toString(),
          collectionId: collection,
          galleryRoot: CONTRACT_CURATION,
          galleryIndex: galleryId.toBigInt(),
          exhibitRoot: CONTRACT_CURATION,
          exhibitIndex: id.toBigInt(),
          source: origin,
          destination: owner,
          amount: price.toString(),
          quantity: "1",
          currency: "0x0000000000000000000000000000000000000000",
          txHash: event.transactionHash,
          txType: "SALE",
          txTimestamp: updatedAt,
        },
        update: {
          txHash: event.transactionHash,
          txType: "SALE",
          txTimestamp: updatedAt,
        },
      });
      txs.push(updateTransaction);
    }

    try {
      const [_, updatedState] = await prisma.$transaction(txs);
      if (updatedState.exhibit_excuted_offset != blockNo) {
        return { success: false, state };
      }
      const newState = {
        ...state,
        exhibit_excuted_offset: blockNo,
      };
      if (eventType == "ExhibitSold" && owner) {
        const key = KEY_COLLECTION_TOKENS(CHAIN, collection);
        const token = await redis.HGET(key, tokenId.toString());
        const parsed = JSON.parse(token);
        parsed["owner"] = owner;
        parsed["ownership"] = [owner];
        await redis.HSET(key, tokenId.toString(), JSON.stringify(parsed));
      }
      return { success: true, state: newState };
    } catch (e) {
      handleError(e);
      return { success: false, state };
    }
  }
}

File: apt/observer/curation/exhibit.py
from datetime import datetime
from typing import List, Tuple
from common.util import new_uuid
from model.token_id import TokenDataId, TokenId
from observer.observer import Observer
from model.curation.exhibit_event import ExhibitEvent, ExhibitEventData
from model.state import State
from model.event import Event
from common.db import prisma_client
from prisma import enums
from config import config

event_type_to_status = {
    "ExhibitListed": enums.CurationExhibitStatus.listing,
    "ExhibitFrozen": enums.CurationExhibitStatus.frozen,
    "ExhibitRedeemed": enums.CurationExhibitStatus.redeemed,
    "ExhibitSold": enums.CurationExhibitStatus.sold
}


class ExhibitEventObserver(Observer[ExhibitEvent]):

    async def process_all(self, state: State, events: List[Event[ExhibitEvent]]) -> State:
        return await super().process_all(state, events)

    async def process(self, state: State, event: Event[ExhibitEvent]) -> Tuple[State, bool]:
        new_state = state
        seqno = event.sequence_number
        data = ExhibitEventData(**event.data)
        gallery_index = int(data.gallery_id)
        event_type = data.event_type
        token_id = TokenId(**data.token_id)
        token_data_id = TokenDataId(**token_id.token_data_id)
        updated_at = datetime.fromtimestamp(int(data.timestamp))
        expired_at = datetime.fromtimestamp(int(data.expiration))

        new_status = event_type_to_status[event_type]
        async with prisma_client.tx(timeout=60000) as transaction:
            result = await transaction.curationexhibit.upsert(
                where={
                    'index_root_status': {
                        'index': int(data.id),
                        'root': config.curation.address(),
                        'status': enums.CurationExhibitStatus.listing
                    }
                },
                data={
                    'create': {
                        'id': new_uuid(),
                        'chain': enums.Chain.APTOS,
                        'index': int(data.id),
                        'root': config.curation.address(),
                        'galleryIndex': gallery_index,
                        'curator':  data.curator,
                        'collectionIdentifier': token_data_id.collection,
                        'tokenIdentifier': token_data_id.name,
                        'tokenCreator': token_data_id.creator,
                        'propertyVersion': int(token_id.property_version),
                        'origin': data.origin,
                        'price': data.price,
                        'currency': '0x1::aptos_coin::AptosCoin',
                        'decimals': 8,
                        'expiredAt': expired_at,
                        'location': data.location,
                        'url': data.url,
                        'detail': data.detail,
                        'status': new_status,
                        'updatedAt': updated_at
                    },
                    'update': {
                        'chain': enums.Chain.APTOS,
                        'index': int(data.id),
                        'root': config.curation.address(),
                        'galleryIndex': gallery_index,
                        'curator':  data.curator,
                        'collectionIdentifier': token_data_id.collection,
                        'tokenIdentifier': token_data_id.name,
                        'tokenCreator': token_data_id.creator,
                        'propertyVersion': int(token_id.property_version),
                        'origin': data.origin,
                        'price': data.price,
                        'currency': '0x1::aptos_coin::AptosCoin',
                        'decimals': 8,
                        'expiredAt': expired_at,
                        'location': data.location,
                        'url': data.url,
                        'detail': data.detail,
                        'status': new_status,
                        'updatedAt': updated_at
                    }
                }
            )
            if result == None or result.status != new_status:
                raise Exception(
                    f'[Curator exhibit]: Failed to update exhibit({data})')

            # token
            updated_token = await transaction.aptostoken.update(
                where={
                    'creator_name_collection_propertyVersion': {
                        'name': token_data_id.name,
                        'creator': token_data_id.creator,
                        'collection': token_data_id.collection,
                        'propertyVersion': '0'
                    }
                },
                data={
                    'owner': data.owner
                }
            )
            if updated_token == None:
                raise Exception(
                    f"[Curator Offer]: Failed to update token owner with {event_type}")

            if result.status == enums.CurationExhibitStatus.sold:
                new_tx = await transaction.transaction.create(
                    data={
                        'id': new_uuid(),
                        'chain': enums.Chain.APTOS,
                        'tokenId': updated_token.id,
                        'collectionId': updated_token.collectionId,
                        'galleryRoot': config.curation.address(),
                        'galleryIndex': gallery_index,
                        'exhibitRoot': config.curation.address(),
                        'exhibitIndex': int(data.id),
                        'source': data.origin,
                        'destination': data.owner,
                        'amount': data.price,
                        'quantity': '1',
                        'currency': '0x1::aptos_coin::AptosCoin',
                        'txHash': f'{event.version}',
                        'txType': enums.TxType.SALE,
                        'txTimestamp': updated_at
                    }
                )
                if new_tx == None:
                    raise Exception(f'[Transaction]: Failed to add new tx')
            updated_offset = await transaction.eventoffset.update(
                where={'id': 0},
                data={
                    "exhibit_excuted_offset": int(seqno)
                }
            )
            if updated_offset == None:
                raise Exception(
                    f'[Curator exhibit]: Failed to update offset')

            new_state.new_offset.exhibit_excuted_offset = updated_offset.exhibit_excuted_offset
            return new_state, True
